### PR TITLE
 gh-121996: Introduce --disable-safety and --enable-slower-safety options

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-121996

----
📚 Documentation preview 📚: https://cpython-previews--122054.org.readthedocs.build/

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -252,6 +252,7 @@ jobs:
         ../cpython-ro-srcdir/configure \
           --config-cache \
           --with-pydebug \
+          --enable-slower-safty \
```
Comment (thesamesam) [inline .github/workflows/build.yml]: s/safty/safety/ everywhere please

File: configure.ac [hunk 2]
```diff
@@ -2458,9 +2458,28 @@ AS_VAR_IF([with_strict_overflow], [yes],
...
+
+if test "$disable_safty" = "no"
+then
+  AX_CHECK_COMPILE_FLAG([-fstack-protector-strong], [BASECFLAGS="$BASECFLAGS -fstack-protector-strong"], [AC_MSG_WARN([-fstack-protector-strong not supported])], [-Werror])
+  AX_CHECK_COMPILE_FLAG([-Wtrampolines], [BASECFLAGS="$BASECFLAGS -Wtrampolines"], [AC_MSG_WARN([-Wtrampolines not supported])], [-Werror])
+fi
+
+AC_MSG_CHECKING([for --enable-slower-safty])
+AC_ARG_ENABLE([slower-safty],
+  [AS_HELP_STRING([--enable-slower-safty], [enable usage of the security compiler options with performance overhead])],[])
```
Comment (thesamesam) [inline configure.ac]: The convention for this in other projects is usually `--with-hardening`.

File: configure.ac [hunk 3]
```diff
@@ -2458,9 +2458,28 @@ AS_VAR_IF([with_strict_overflow], [yes],
...
+AC_MSG_CHECKING([for --disable-safty])
+AC_ARG_ENABLE([safty],
+  [AS_HELP_STRING([--disable-safty], [disable usage of the security compiler options with no performance overhead])],
+  [AS_VAR_IF([enable_safty], [yes], [disable_safty=no], [disable_safty=yes])], [disable_safty=no])
+AC_MSG_RESULT([$disable_safty])
+
+if test "$disable_safty" = "no"
+then
+  AX_CHECK_COMPILE_FLAG([-fstack-protector-strong], [BASECFLAGS="$BASECFLAGS -fstack-protector-strong"], [AC_MSG_WARN([-fstack-protector-strong not supported])], [-Werror])
+  AX_CHECK_COMPILE_FLAG([-Wtrampolines], [BASECFLAGS="$BASECFLAGS -Wtrampolines"], [AC_MSG_WARN([-Wtrampolines not supported])], [-Werror])
```
Comment (thesamesam) [inline configure.ac:2512]: The mixing of warning options with an actual codegen option (SSP) doesn't feel ideal.

Comment (thesamesam) [reply configure.ac:2512]: Also, tbh, pretty much every Linux distro builds with SSP and _F_S=2 at least..

File: configure.ac [hunk 4]
```diff
@@ -2458,9 +2458,28 @@ AS_VAR_IF([with_strict_overflow], [yes],
...
+fi
+
+AC_MSG_CHECKING([for --enable-slower-safty])
+AC_ARG_ENABLE([slower-safty],
+  [AS_HELP_STRING([--enable-slower-safty], [enable usage of the security compiler options with performance overhead])],[])
+AC_MSG_RESULT([$enable_slower_safty])
+
+if test "$enable_slower_safty" = "yes"
+then
+  AX_CHECK_COMPILE_FLAG([-D_FORTIFY_SOURCE=3], [BASECFLAGS="$BASECFLAGS -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3"], [AC_MSG_WARN([-D_FORTIFY_SOURCE=3 not supported])])
```
Comment (thesamesam) [inline configure.ac:2522]: Please use [AX_ADD_FORTIFY_SOURCE](https://www.gnu.org/software/autoconf-archive/ax_add_fortify_source.html).

**Comment by thesamesam** [comment]
It's not clear to me how this PR interacts with https://github.com/python/cpython/pull/121997.

**Comment by corona10** [comment]
> It's not clear to me how this PR interacts with https://github.com/python/cpython/pull/121997.

It's not interact just separate suggestion.

File: configure.ac (hunk 3 above)
Comment (corona10) [reply configure.ac:2512]: Let's handle it as the seperate PR cc  @nohlson

File: configure.ac (hunk 4 above)
Comment (corona10) [reply configure.ac:2522]: https://github.com/python/cpython/pull/122054#discussion_r1686368136 ditto,
This PR is not about compiler options but more about how we can enable them and disable them without interuptting @nohlson 's experimentation.

File: configure.ac [hunk 5]
```diff
@@ -2499,9 +2499,28 @@ AS_VAR_IF([with_strict_overflow], [yes],
...
 # Enable flags that warn and protect for potential security vulnerabilities.
 # These flags should be enabled by default for all builds.
-AX_CHECK_COMPILE_FLAG([-fstack-protector-strong], [BASECFLAGS="$BASECFLAGS -fstack-protector-strong"], [AC_MSG_WARN([-fstack-protector-strong not supported])], [-Werror])
-AX_CHECK_COMPILE_FLAG([-Wtrampolines], [BASECFLAGS="$BASECFLAGS -Wtrampolines"], [AC_MSG_WARN([-Wtrampolines not supported])], [-Werror])
-AX_CHECK_COMPILE_FLAG([-D_FORTIFY_SOURCE=3], [BASECFLAGS="$BASECFLAGS -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3"], [AC_MSG_WARN([-D_FORTIFY_SOURCE=3 not supported])])
+
+AC_MSG_CHECKING([for --disable-safety])
+AC_ARG_ENABLE([safety],
+  [AS_HELP_STRING([--disable-safety], [disable usage of the security compiler options with no performance overhead])],
+  [AS_VAR_IF([enable_safety], [yes], [disable_safety=no], [disable_saftey=yes])], [disable_saftey=no])
```
Comment (kulikjak) [inline configure.ac:2506]: typo: `disable_saftey` should be `disable_safety`

**Pull request merged**